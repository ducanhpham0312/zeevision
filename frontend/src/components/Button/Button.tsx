import React, { ButtonHTMLAttributes } from "react";
import { twMerge } from "tailwind-merge";
import { tv, type VariantProps } from "tailwind-variants";

export interface ButtonProps
  extends ButtonHTMLAttributes<HTMLButtonElement>,
    VariantProps<typeof buttonVariant> {
  width?: number;
  children?: React.ReactNode;
  active?: boolean;
  label?: string;
  helperText?: string;
}

export function Button({
  children,
  width,
  active,
  className,
  variant,
  size,
  fullWidth,
  label,
  helperText,
  ...props
}: ButtonProps) {
  return (
    <button
      {...props}
      className={twMerge(
        buttonVariant({ variant, size, fullWidth }),
        className,
      )}
      style={{ width: `${width}px` }}
    >
      {helperText ? (
        <div className="absolute -left-3 top-[50%] -translate-x-[100%] -translate-y-[50%] scale-50 rounded bg-[#202020] px-2 py-1 text-white opacity-0 transition group-hover:scale-100 group-hover:opacity-100">
          <p className="whitespace-nowrap text-sm">{helperText}</p>
        </div>
      ) : null}
      <div
        className={`transition-width absolute bottom-0 left-0 w-full transform rounded bg-accent duration-100`}
        style={{ height: active ? "4px" : "0px" }}
      />
      <span>{label || children}</span>
    </button>
  );
}

const buttonVariant = tv({
  base: "group hover:shadow-lg transition relative box-border text-accent hover:bg-hover active:bg-active rounded",
  variants: {
    variant: {
      primary: "text-white bg-accent hover:bg-accent/90 active:bg-accent/80",
      secondary: "bg-second-accent",
    },
    size: {
      compact: "px-2 py-1",
      standard: "px-3 py-2",
      large: "px-4 py-4",
    },
    fullWidth: {
      true: "w-full",
    },
  },
  defaultVariants: {
    size: "standard",
  },
});
