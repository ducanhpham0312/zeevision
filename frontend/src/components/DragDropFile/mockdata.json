{
  "file": {
    "content": "Test File DragnDrop",
    "name": "test.bpmn",
    "type": "text/plain"
  }
}
