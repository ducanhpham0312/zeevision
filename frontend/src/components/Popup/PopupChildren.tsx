import { styled } from "@mui/system";

export const PopupContent = styled("div")`
  margin-top: 10px;
  flex-grow: 1;
  padding-left: 20px;
  overflow-y: scroll;
  max-height: 500px;
  > div {
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin-right: 20px;
  }
`;

export const PopupAction = styled("div")`
  display: flex;
  gap: 10px;
  justify-content: flex-end;
  padding: 10px;
`;
