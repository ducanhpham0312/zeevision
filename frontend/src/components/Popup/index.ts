export { Popup } from "./Popup";
export { PopupAction, PopupContent } from "./PopupChildren";
