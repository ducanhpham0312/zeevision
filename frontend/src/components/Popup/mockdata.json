{
  "openPopup": {
    "open": true,
    "title": "Test Title",
    "children": ["Test child 1", "Test child 2"]
  },
  "closedPopup": {
    "open": false,
    "title": "Test Title",
    "children": ["Test child 1", "Test child 2"]
  },
  "primary": {
    "open": true,
    "title": "Deploy a Process",
    "children": []
  },
  "scrollable": {
    "open": true,
    "title": "Deploy a Process",
    "children": [
      "Cras mattis consectetur purus sit amet fermentum.",
      "Cras justo odio, dapibus ac facilisis in, egestas eget quam.",
      "Morbi leo risus, porta ac consectetur ac, vestibulum at eros.",
      "Praesent commodo cursus magna, vel scelerisque nisl consectetur et."
    ]
  }
}
