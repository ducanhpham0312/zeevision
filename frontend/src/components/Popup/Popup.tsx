import { Modal, ModalProps } from "@mui/base";
import CloseIcon from "@mui/icons-material/Close";
import { styled } from "@mui/system";
import { animated, useSpring } from "@react-spring/web";
import React, { MouseEventHandler, forwardRef } from "react";
import { Button } from "../Button";

export interface PopupProps extends Omit<ModalProps, "children"> {
  title: string;
  children: React.ReactNode[];
}

export const Popup: React.FC<PopupProps> = ({ title, children, ...props }) => {
  return (
    <div>
      <StyledModal
        {...props}
        slots={{ backdrop: StyledBackdrop }}
        closeAfterTransition
      >
        <Fade in={props.open}>
          <PopupContainer>
            <PopupTitle>
              <p>{title}</p>
              <Button variant={"text"} onClick={props.onClose as () => void}>
                <CloseIcon />
              </Button>
            </PopupTitle>
            {children}
          </PopupContainer>
        </Fade>
      </StyledModal>
    </div>
  );
};

interface FadeProps {
  children: React.ReactElement;
  in?: boolean;
  onClick?: MouseEventHandler<HTMLDivElement> | undefined;
  onEnter?: (node: HTMLElement, isAppearing: boolean) => void;
  onExited?: (node: HTMLElement, isAppearing: boolean) => void;
}

const Fade = forwardRef<HTMLDivElement, FadeProps>(function Fade(props, ref) {
  const { in: open, children, onEnter, onExited, ...other } = props;
  const style = useSpring({
    from: { opacity: 0 },
    to: { opacity: open ? 1 : 0 },
    config: { duration: 150 },
    onStart: () => {
      if (open && onEnter) {
        onEnter(null as unknown as HTMLElement, true);
      }
    },
    onRest: () => {
      if (!open && onExited) {
        onExited(null as unknown as HTMLElement, true);
      }
    },
  });
  return (
    <animated.div ref={ref} style={style} {...other}>
      {children}
    </animated.div>
  );
});

const Backdrop = forwardRef<
  HTMLDivElement,
  { children: React.ReactElement; open: boolean }
>((props, ref) => {
  const { open, ...other } = props;
  return <Fade ref={ref} in={open} {...other} />;
});

export const PopupTitle = styled("div")`
  display: flex;
  height: 50px;
  line-height: 5px;
  justify-content: space-between;
  padding: 10px;
  padding-left: 20px;

  p {
    font-size: 25px;
    font-weight: 700;
  }
`;

export const PopupContent = styled("div")`
  margin-top: 10px;
  flex-grow: 1;
  padding-left: 20px;
  overflow-y: scroll;
  max-height: 500px;
  > div {
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin-right: 20px;
  }
`;

export const PopupAction = styled("div")`
  display: flex;
  gap: 10px;
  justify-content: flex-end;
  padding: 10px;
`;

export const StyledBackdrop = styled(Backdrop)`
  z-index: -1;
  position: fixed;
  inset: 0;
  background-color: rgb(0 0 0 / 0.5);
  -webkit-tap-highlight-color: transparent;
`;

// make this responsive
export const PopupContainer = styled("div")`
  width: 800px;
  border: 1px solid black;
  overflow-y: auto;
  max-height: 90%;
  min-height: 300px;
  display: flex;
  flex-direction: column;
  background-color: white;
  border-radius: 8px;
`;

export const StyledModal = styled(Modal)`
  position: fixed;
  z-index: 1300;
  inset: 0;
  display: flex;
  align-items: center;
  justify-content: center;
`;
