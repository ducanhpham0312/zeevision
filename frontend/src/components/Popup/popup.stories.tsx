import type { Meta, StoryObj } from "@storybook/react";
import { Button } from "../Button";
import { Popup, PopupAction, PopupContent } from "./Popup";
import { useState } from "react";
import mockdata from "./mockdata.json";

const meta = {
  title: "Assets/Popup",
  component: Popup,
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
  argTypes: {},
} satisfies Meta<typeof Popup>;

export default meta;
type Story = StoryObj<typeof meta>;

const ButtonWithHooks = ({ text }: { text?: Array<string> }) => {
  const [open, setOpen] = useState(false);
  const handleOpen = () => setOpen(true);
  const handleClose = () => setOpen(false);
  return (
    <>
      <Button onClick={handleOpen}>Open Modal</Button>
      <Popup open={open} onClose={handleClose} title={"Deploy a process"}>
        <PopupContent>
          <div>
            <Button>Deploy a file (.bpmn) </Button>
            <p>Or drag the files to the box belows</p>
          </div>
          {text && (
            <div>{[...Array(50)].map(() => text.join("\n"))}</div>
          )}
        </PopupContent>
        <PopupAction>
          <Button onClick={handleClose}>Cancel</Button>
          <Button variant="contained">Deploy process</Button>
        </PopupAction>
      </Popup>
    </>
  );
};

export const Primary: Story = {
  args: {
    open: mockdata.primary.open,
    title: mockdata.primary.title,
    children: mockdata.primary.children,
  },
  render: () => <ButtonWithHooks />,
};

export const Scrollable: Story = {
  args: {
    open: mockdata.scrollable.open,
    title: mockdata.scrollable.title,
    children: mockdata.scrollable.children,
  },
  render: () => <ButtonWithHooks text={mockdata.scrollable.children} />,
};
