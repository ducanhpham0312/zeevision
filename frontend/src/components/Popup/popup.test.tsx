import { render, fireEvent, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { Popup } from "./Popup";
import { Button } from "../Button";
import * as mockdata from "./mockdata.json";

describe("Popup Component", () => {
  const mockOnClose = jest.fn();

  it("renders correctly when open", () => {
    const { baseElement } = render(
      <Popup {...mockdata.openPopup} onClose={mockOnClose} />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it("does not render when closed", () => {
    const { baseElement } = render(
      <Popup {...mockdata.closedPopup} onClose={mockOnClose} />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it("calls onClose when close button is clicked", () => {
    const onClose = jest.fn();
    render(<Popup {...mockdata.openPopup} onClose={onClose} />);
    fireEvent.click(screen.getByRole("button"));
    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it("renders the title correctly", () => {
    render(<Popup {...mockdata.openPopup} onClose={mockOnClose} />);
    expect(screen.getByText("Test Title")).toBeInTheDocument();
  });

  it("renders children correctly", () => {
    render(<Popup {...mockdata.openPopup} onClose={mockOnClose} />);
    expect(screen.getByText(/Test child 1/)).toBeInTheDocument();
    expect(screen.getByText(/Test child 2/)).toBeInTheDocument();
  });

  it("renders a button with correct text and form of Button Component inside the Popup", () => {
    const buttonText = "Click Me!";
    render(
      <Popup {...mockdata.openPopup}>
        <p>Test child 1</p>
        <p>Test child 2</p>
        <Button>{buttonText}</Button>
      </Popup>
    );

    expect(screen.getByText(buttonText)).toBeInTheDocument();
  });

  it("Backdrop displays correctly when the popup is open", () => {
    render(<Popup {...mockdata.openPopup} onClose={mockOnClose} />);
    const backdrop = screen.getByRole("presentation");
    expect(backdrop).toHaveStyle({
      position: "fixed",
      inset: "0",
      backgroundColor: "rgb(0 0 0 / 0.5)",
    });
  });
});
