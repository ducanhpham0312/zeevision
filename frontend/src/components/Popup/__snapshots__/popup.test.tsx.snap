// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup Component does not render when closed 1`] = `
<body
  style=""
>
  <div>
    <div />
  </div>
</body>
`;

exports[`Popup Component renders correctly when open 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div />
  </div>
  <div
    class="MuiModal-root fixed inset-0 z-50 flex w-full items-center justify-center"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiModal-backdrop css-1jyxheb"
      style="opacity: 0;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      style="opacity: 0;"
      tabindex="-1"
    >
      <div
        class="flex max-h-[90%] min-h-[300px] w-[90vw] flex-col overflow-y-auto rounded bg-background tablet:w-[750px]"
      >
        <div
          class="flex items-center justify-between px-[20px] py-[10px] pl-[20px] text-2xl"
        >
          <p>
            Test Title
          </p>
          <button
            class="relative box-border text-accent hover:bg-hover active:bg-active rounded px-3 py-2"
          >
            <div
              class="transition-width absolute bottom-0 left-0 w-full transform rounded bg-accent duration-100"
              style="height: 0px;"
            />
            <span>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CloseIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </span>
          </button>
        </div>
        Test child 1
        Test child 2
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
